phoenix-core/src/main/java/org/apache/phoenix/compile/ScanRanges.java

```diff
@@ -694,4 +694,136 @@ public TimeRange getRowTimestampRange() {
+                    byte[] lowerBound;
+                    byte[] upperBound;
+
+                    // Note ScanUtil.setKey internally handles the upper/lower exclusive from a Scan
+                    // point of view. It would be good to break it out in the future for rowkey
+                    // construction vs ScanKey construction To handle differences between
+                    // hbase 2 and hbase 1 scan boundaries
+                    int result = ScanUtil.setKey(schema, newRanges, slotSpans, positions, KeyRange.Bound.LOWER, keyBuffer,
+                            0, 0, slotSpans.length);
+
+                    if (result < 0) {
+                        // unbound
+                        lowerBound = KeyRange.UNBOUND;
+                    } else {
+                        lowerBound = Arrays.copyOf(keyBuffer, result);
+                    }
+
+                    result = ScanUtil.setKey(schema, newRanges, slotSpans, positions, KeyRange.Bound.UPPER, keyBuffer, 0, 0,
+                            slotSpans.length);
+
+                    if (result < 0) {
+                        // unbound
+                        upperBound = KeyRange.UNBOUND;
+                    } else {
+                        upperBound = Arrays.copyOf(keyBuffer, result);
+                    }
+
+
+                    /*
+                     * This is already considered inside of ScanUtil.setKey we may want to refactor to pull these out.
+                     * range/single    boundary       bound      increment
+                     *  range          inclusive      lower         no
+                     *  range          inclusive      upper         yes, at the end if occurs at any slots.
+                     *  range          exclusive      lower         yes
+                     *  range          exclusive      upper         no
+                     *  single         inclusive      lower         no
+                     *  single         inclusive      upper         yes, at the end if it is the last slots.
+                     */
+                    
+                    boolean lowerInclusive = true;
+                    boolean upperInclusive = true;
+                    // If we are not a single key
+                    if (!Arrays.equals(lowerBound, upperBound)) {
+                        upperInclusive = false;
+                    }
+
+                    // Don't send a null range, send an empty range.
+                    if (lowerBound.length == 0 && upperBound.length == 0) {
+                        lowerInclusive = false;
+                        upperInclusive = false;
+                    }
+
+                    KeyRange keyRange = KeyRange.getKeyRange(lowerBound, lowerInclusive, upperBound, upperInclusive);
+                    expandedRanges.add(keyRange);
+
+                    // update position
+                    int i;
+                    for (i = positions.length - 1; i >= 0; i--) {
```

Comment:
Can you explain the usage of this loop?

Reply:
Added additional comments but this is a counting loop from the trailing edge to the leading edge for all possible settings of the Keys based on the ranges.  That is for PK (key1,key2) and Where clause key1 in {a,b,c} and key2 in {x,y,z}, then we need to generate keys ax,ay,az,bx,by,bz,cx,cy,cz.  See the setKey method for more details.